File: src/com/pytest_support/fixtures/FixtureNamesCompletionContributor.java
package com.pytest_support.fixtures;

import com.intellij.codeInsight.completion.*;
import com.intellij.codeInsight.lookup.LookupElementBuilder;
import com.intellij.util.ProcessingContext;
import com.jetbrains.python.psi.PyFile;
import com.jetbrains.python.psi.PyFunction;
import com.pytest_support.consts.PytestConsts;
import com.pytest_support.utils.FixtureUtils;
import com.pytest_support.utils.StreamUtils;
import icons.PythonIcons;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;

public class FixtureNamesCompletionContributor extends CompletionContributor {
    public FixtureNamesCompletionContributor() {
        this.extend(
                CompletionType.BASIC,
                PytestConsts.functionParameter,
                new FixtureCompletionProvider()
        );
    }

    private static class FixtureCompletionProvider extends CompletionProvider<CompletionParameters> {

        @Override
        protected void addCompletions(
                @NotNull CompletionParameters completionParameters,
                ProcessingContext processingContext,
                @NotNull CompletionResultSet result) {

            if (FixtureUtils.isTestOrFixtureParameter(completionParameters.getPosition())) {
                result.addElement(LookupElementBuilder.create("request").withIcon(PythonIcons.Python.Function));
                StreamUtils.pyFixturesStream((PyFile) completionParameters.getOriginalFile())
                        .filter(FixtureUtils::isFixture)
                        .map(PyFunction::getName)
                        .filter(Objects::nonNull)
                        .forEach(name -> result.addElement(
                                LookupElementBuilder.create(name).withIcon(PythonIcons.Python.Function)
                        ));
                result.stopHere();
            }
        }
    }
}

File: src/com/pytest_support/fixtures/FixtureNamesCompletionContributor.kt
package com.pytest_support.fixtures

import com.intellij.codeInsight.completion.*
import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.intellij.util.ProcessingContext
import com.jetbrains.python.psi.PyFile
import com.pytest_support.consts.functionParameter
import com.pytest_support.extensions.fixtures
import com.pytest_support.extensions.isTestOrFixtureParameter
import com.pytest_support.extensions.names
import icons.PythonIcons


val REQUEST_ELEMENT: LookupElementBuilder = LookupElementBuilder
        .create("request")
        .withIcon(PythonIcons.Python.Function)


class FixtureNamesCompletionContributor : CompletionContributor() {
    init {
        this.extend(
                CompletionType.BASIC,
                functionParameter,
                FixtureCompletionProvider()
        )
    }

    private class FixtureCompletionProvider : CompletionProvider<CompletionParameters>() {

        override fun addCompletions(
                completionParameters: CompletionParameters,
                processingContext: ProcessingContext,
                result: CompletionResultSet
        ) {
            if (completionParameters.position.isTestOrFixtureParameter()) {
                result.addElement(REQUEST_ELEMENT)

                (completionParameters.originalFile as PyFile)
                        .fixtures()
                        .names()
                        .map {
                            LookupElementBuilder
                                    .create(it)
                                    .withIcon(PythonIcons.Python.Function)
                        }
                        .forEach { result.addElement(it) }

                result.stopHere()
            }
        }
    }
}
